src/Biogeochemistry.jl

```diff
@@ -0,0 +1,134 @@
+module Biogeochemistry
+using Oceananigans.Grids: Center, xnode, ynode, znode
+using Oceananigans.Forcings: maybe_constant_field, DiscreteForcing
+using Oceananigans.Advection: div_Uc, UpwindBiasedFifthOrder
+
+import Oceananigans.Fields: location
+import Oceananigans.Forcings: regularize_forcing
+#####
+##### Generic fallbacks for biogeochemistry
+#####
+
+nothing_function(args...) = nothing
+
+"""
+Update tracer tendencies.
+
+Called at the end of calculate_tendencies!
+"""
+update_tendencies!(bgc, model) = nothing
+
+"""
+Update tracer tendencies.
+
+Called at the end of calculate_tendencies!
+"""
+update_biogeochemical_state!(bgc, model) = nothing
+
+"""Return the biogeochemical forcing for `val_tracer_name` when model is called."""
+abstract type AbstractBiogeochemistry end
+
+@inline (::AbstractBiogeochemistry)(i, j, k, grid, val_tracer_name, clock, fields) = zero(grid)
+
+struct NoBiogeochemistry <: AbstractBiogeochemistry end
+
+tracernames(tracers) = keys(tracers)
+tracernames(tracers::Tuple) = tracers
+
+"""Ensure that `tracers` contains biogeochemical tracers and `auxiliary_fields` contains biogeochemical auxiliary fields (e.g. PAR)."""
+@inline function validate_biogeochemistry!(tracers, auxiliary_fields, bgc)
+    req_tracers = required_biogeochemical_tracers(bgc)
+    
+    all(tracer ∈ tracernames(tracers) for tracer in req_tracers) ||
+        error("$(req_tracers) must be among the list of tracers to use $(typeof(bgc).name.wrapper)")
+
+    req_auxiliary_fields = required_biogeochemical_auxiliary_fields(bgc)
+    
+    all(field ∈ tracernames(auxiliary_fields) for field in req_auxiliary_fields) ||
+        error("$(req_auxiliary_fields) must be among the list of auxiliary fields to use $(typeof(bgc).name.wrapper)")
+    
+    return nothing
+end
+
+required_biogeochemical_tracers(::NoBiogeochemistry) = ()
+required_biogeochemical_auxiliary_fields(bgc::AbstractBiogeochemistry) = ()
+
+"""Sets up a tracer based biogeochemical model in a similar way to SeawaterBouyancy"""
+struct TracerBasedBiogeochemistry <: AbstractBiogeochemistry
```

Comment:
Is this the right name?

Reply:
Maybe something just like `BiogeochemicalModel` would be better?